### DISCUSSION TITLE
Why not using unix domain sockets instead of uwsgi

### DISCUSSION THREAD

**Discussion by cyclaero** [searx/searx#2591] [General]
I found out that `webapp.py` may be directly bound to a unix domain socket. By `settings.yml` I passed `bind_address : "unix:///tmp/searx.sock“`. No need to create a `searx` user, here I run the searx instance as the `www`user.

For Apache I enabled mod_proxy and mod_proxy_http, and created a virtual host entry for my searx domain. The second entry allows me to use my searx instance instead of bing in the address bar of Safari — this is optional of course.

    <VirtualHost *:80>
       ServerName            searx.example.com:80
       ProxyPass             / unix:/tmp/searx.sock|http://searx.example.com/
    </VirtualHost>

    #<VirtualHost *:443>
    #   ServerName            www.bing.com:443
    #   RedirectPermanent     / http://searx.example.com/
    #
    #   SSLEngine             on
    #   SSLCertificateFile    "/usr/local/etc/squid/proxy-certs/proxy-ca.crt"
    #   SSLCertificateKeyFile "/usr/local/etc/squid/proxy-certs/proxy-ca.pem"
    #</VirtualHost>

**Comment by return42** [comment]
In your example there is no bot blocker what is OK for a intranet solution. For internet solutions we recommend to use a bot blocker in front of searx like shown in this [architecture figure](https://searx.github.io/searx/admin/architecture.html).  In this figure you can see filtron in front of searx, I guess filtron can't speak to unix socest.  In this figure you can also see, that uwsgi is a layer to multiplex. unix sockets are processing request sequential.

**Reply by cyclaero** [reply]
Yes, I am talking about searx instances in the LAN only. Here, I installed searx along side a web server which is serving the public internet, and in order to prevent access from outside to searx, I added the following to it's virtual host configuration:

    <Proxy *>
       Require   ip 192.168.0.0/24
    </Proxy>

PS:
Usually unix domain sockets are much more performant than IP sockets, and at the same time offer the security benefit of being absolutely local only. This is why we want to use UDS instead of IPS wherever possible. If the other layers which you put in front of searx are not able to communicate over UDS, then of course this is a no go.

https://lists.freebsd.org/pipermail/freebsd-performance/2005-February/001143.html

**Reply by return42** [reply]
> Usually unix domain sockets are much more performant than IP sockets.

They do not have a TCP/IP overlay, that makes them a bit faster. But they can't process requests in parallel.  If you are the only user of the service that might be OK for you :)

BTW: `localhost` and `127.0.0.1`  _"offer the security benefit of being absolutely local only."_ ;)

**Reply by cyclaero** [reply]
> ... But they can't process requests in parallel.

Since when?

**Reply by return42** [reply]
> Since when?

Ask your self; how many process do listen on a single socket?

**Reply by cyclaero** [reply]
I programmed already server daemons using both kinds of sockets, namely UDS (Unix Domain Sockets) and IPS (Internet Protocol Sockets) in C for FreeBSD which run out of the box on any *BSD system and on macOS, and with minor modifications would run on Linux as well -- one example is this: https://github.com/cyclaero/ContentCGI

When it comes to the socket layer, the behaviour and API of both are identical up to the extend of the present concern. The mother process may bind() to one endpoint and then fork() off child processes. The childs share automatically all open file descriptors and may then all listen() to the one and only socket which was bounded by the mother process. The kernel maintains a FIFO with all incoming connection requests to a given bounded socket (either of UDS or IPS) and passes theses in a first-in first-out manner to respective accept() calls by any of the child and/or mother processes. These do then the actual work, usually in a pthread in order to be able to listen() immediately again after pthread_create(). Therefore, the requests may be processed in parallel while new requests are coming in. And still to this point there is no difference between UDS and IPS. Both sockets can be used by multiple processes (sharing the same mother) and each having any number of threads limited only by the system's resources.

So yes, UDS can serve requests in parallel exactly like IPS and in addition it is an order of magnitude faster, because it does not have that much overhead.

**Reply by return42** [reply]
@cyclaero thanks a lot for the detailed explanation, I did not know theses details about sockets / very interesting for me. There is one question I have .. 

> I found out that webapp.py may be directly bound to a unix domain socket. 

Did you start multiple `webapp.py` process listening on `bind_address : "unix:///tmp/searx.sock“`?

**Reply by cyclaero** [reply]
> Did you start multiple `webapp.py` process listening on `bind_address : "unix:///tmp/searx.sock“`?

No, here I use a single instance in our LAN only. Anyway, it is faced by Apache which runs in threaded multiple worker mode, and I won’t expect much congestion problems even with some higher workload. However, it can be done, but not as simple as launching several `webapp.py` instances by any shell, in this case only the last launched one would receive all connection requests. We need to launch a mother process, which may be but not necessarily needs to be a `webapp.py` instance and which does one `bind()` for all. Then this may `fork()` off other `webapp.py` child processes which must not call `bind()` by itself but instead need to use the already open file descriptor from the mother process. From there on, everything else, i.e. `listen()` and the subsequent `accept()` - `pthread_create()` - `recv()`- `send()` loop would be all the same.

I see no reason why this won’t work. However, minor changes to `webapp.py` would be necessary. In the mother `webapp.py` you would need to add a loop where it forks off a configured number of childs, and it needs to pass the socket file descriptor which resulted from its call to `bind()` somehow to these forks. This could simply be done by an `environ` variable, for example `bind_socket_fd`. In the child `webapp.py` you need to add an `if ()` statement which calls `bind()` and enters the `fork()` off loop only when the `bind_socket_fd` variable does not exist. Otherwise it would straight go to the `listen()` step.

I am not safe in Python, however in C, forking off child processe does usually take 5 lines of code, and said `if ()` statement should be even less. Somebody who knows the code of `webapp.py`could do this quite quickly.

**Reply by cyclaero** [reply]
Well, I did some tests with with multiple instances of the SearX `webapp.py` as is on FreeBSD 12.2-RELEASE-p3.

Results:

More than one instance won’t bind on 127.0.0.1:8888, for the second launch I see:

     Traceback (most recent call last):
     File "/usr/local/lib/python3.7/site-packages/searx/webapp.py", line 1156, in <module>
       run()
     File "/usr/local/lib/python3.7/site-packages/searx/webapp.py", line 1110, in run
       threaded=True
     File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 990, in run
       run_simple(host, port, self, **options)
     File "/usr/local/lib/python3.7/site-packages/werkzeug/serving.py", line 1052, in run_simple
       inner()
     File "/usr/local/lib/python3.7/site-packages/werkzeug/serving.py", line 1005, in inner
       fd=fd,
     File "/usr/local/lib/python3.7/site-packages/werkzeug/serving.py", line 848, in make_server
       host, port, app, request_handler, passthrough_errors, ssl_context, fd=fd
     File "/usr/local/lib/python3.7/site-packages/werkzeug/serving.py", line 740, in __init__
       HTTPServer.__init__(self, server_address, handler)
     File "/usr/local/lib/python3.7/socketserver.py", line 452, in __init__
       self.server_bind()
     File "/usr/local/lib/python3.7/http/server.py", line 137, in server_bind
       socketserver.TCPServer.server_bind(self)
     File "/usr/local/lib/python3.7/socketserver.py", line 466, in server_bind
       self.socket.bind(self.server_address)
     OSError: [Errno 48] Address already in use

I am able to launch any number of SearX `webapp.py` instances binding and listening to the same UDS (I tried 5 in separate terminal windows). For this test I manually launched the 5 using:
`sudo -u www python3.7 /usr/local/lib/python3.7/site-packages/searx/webapp.py`.

To my great surprise, all of them are actually listening behind the very same Apache proxy, however the last launched one is the most active and the first launched one receives requests only under very heavy load, which was hard to produce in my setup.

When I stop one of them, then all others stop listening as well, probably because the UDS becomes unlinked upon exit of any one `webapp.py`.
